### ISSUE TITLE
Add ZstdDictInputStream and ZstdDictOutputStream

### ISSUE THREAD

**Issue by KennethLaw** [luben/zstd-jni#33]
Can you add useful InputStream class for dict compress/decompress?

**Comment by luben** [comment]
I have intention to do so but I am not sure it's going to be very useful: dictionary compression makes sense only for very small objects and in this case there are going to fit into one zstd buffer in 128K. If you use dict compression over bigger chunks it will not give you measurable benefit.

Still it's good to have in order to keep all Zstd capabilities exposed.

I will take a look after zstd-1.1.5 release as 1.1.4 have some inconsistency between creating compression/decompression dictionaries and the work around is too hacky for my taste. It's already fixed in the dev branch of zstd so I expect it to have fixed release in less than a month.

**Comment by hari-rangarajan** [comment]
Wondering if this is still on your roadmap?

**Comment by hari-rangarajan** [comment]
@luben On the usefulness question: Smaller objects can be parsed (e.g. json) directly into the final representation without the need for an intermediate object using the streaming interface.

**Comment by luben** [comment]
Hi,

It's still on my road map but not as highest priority. There is a new configuration API for Zstd that is still experimental that should make this change easy but we have to port the current code to it. I am waiting on it to be declared stable to switch to it and most probably the streaming compression/decompression with dictionaries will come with it.

**Comment by scottcarey** [comment]
As for the usefulness of it: https://amplitude.engineering/reducing-kafka-costs-with-z-tandard-88e0e489850f

Its pretty handy with block compression over large/infinite data streams.

**Comment by dongjinleekr** [comment]
@luben Hello. I am applying your wonderful library into Apache Kafka: apache/kafka#2267 As @scottcarey  already mentioned, this feature will be useful when compressing/decompressing large/infinite data streams. If it is still on your roadmap, may I take this issue? I already have some experience in implementing Java bindings.

**Comment by luben** [comment]
Hi, you can take a shot on the problem, recently I didn't have a lot of time to work on it.

I have a question how it is going to work if there are multiple publishers - Kafka stream is not limited to just one publisher. So if 2 publishers start pushing events they can't use the same compression stream, as the result depends on the previous state/stream. Or is the compression is going to happen server side?

Thanks in advance,
luben

**Comment by scottcarey** [comment]
The stream can be compressed client side too.

Some ideas:

The block would probably be of the format:  { dictionaryId, compressedData }  or otherwise contain which dictionary was used when compressing.  A map of active dictionaryIds -> dictionaryBytes would have to be kept by the broker per topic (or partition), and clients would not be allowed to compress a block without a known dictionary (or no dictionary).    The active ones could change over time.

The input stream here would be per block, not infinite itself, but blocks can be large enough that we don't want to inflate the whole thing into a ByteBuffer.  One can construct an infinite InputStream from a chain of smaller InputStreams, if one wanted to do that.

**Comment by dongjinleekr** [comment]
@scottcarey I greatly appreciate your opinion. I am designing the zstd-dictionary support in Kafka, analyzing the other completed implementation. At this point, the blueprint in my mind is almost identical to yours.

However, the zstd compression in Kafka and dictionary feature in Zstd are the prerequisites, let's concentrate on those issues. When I have to write the KIP of this feature, I will invite you and @luben to the conversation. Stay tuned!

**Comment by luben** [comment]
@dongjinleekr if you hit any roadblocks with non-dict zstd integration, don't hesitate to contact me.

**Comment by luben** [comment]
In 1aae8f4d I have added support for byte array dictionaries in the Input/Output streams, also to the variants working on ByteBuffers.

Before release I want to add support for CDicts and write some tests. Please build it from source for testing. Any feedback is welcome!

**Comment by luben** [comment]
zstd-jni-1.3.5-2 with fill support for dictionaries in the streaming modes in on the way to maven central. This should unblock  apache/kafka#2267 for exploring dictionary compression.

**Comment by luben** [comment]
Closing, please file an new issue if you have any problem with the new functionality
